Add support for stm32h7 dma driver

shlomow opened the pull request:
This pr adds support for stm32h7 socs in the dma driver.
Only changes to how dmamux work along dma v1 and not only v2 were needed.
Add also nucleo-h743zi to uart async api test

shlomow committed:
dts: stm32: h7: add missing dma properties

Add dma-offset and dma-requests missing
device tree properties.

Signed-off-by: Shlomi Vaknin <shlomi.39sd@gmail.com>

FRASTM commented on drivers/dma/dma_stm32.h:
```diff
@@ -44,17 +44,17 @@ struct dma_stm32_config {
 	struct dma_stm32_stream *streams;
 };
 
-#ifdef CONFIG_DMA_STM32_V1
```
With stm32H7, DMA V1 has a DMAMUX (not only DMA V2). However for soc with DMA V2 and NO DMAMUX, the _stream id is in range 1..<dma-requests>_

erwango replied:
Is that resolved ?

shlomow replied:
Yes. Now, the stream offset is 1, if we use dmamux or dma_v2. In the past, the offset was 0 if we dma_v1 which is not true for h7 socs. By the way, in the future it needs to be refactored in order to add `bdma` support which is dma_v2 controller. So a way to use dma_v1 in parallel to dma_v2 should be possible

FRASTM commented:
This configuration of DMAMUX multiplexing a DMA V1 is new, could you please confirm that tests/drivers/dma and tests/drivers/spi/spi_loopback are passed on this nucleo_h743 board.

shlomow committed:
drivers: dma: stm32: add support for stm32h7

Add the missing parts for adding support
to stm32h7 dma driver.
The fix is to make dmamux driver work with
dma v1 driver.

Signed-off-by: Shlomi Vaknin <shlomi.39sd@gmail.com>

shlomow committed:
tests: dma: add stm32h7 to chan_blen_transfer

Add nucleo h743zi board to chan_blen_transfer test.

Signed-off-by: Shlomi Vaknin <shlomi.39sd@gmail.com>

shlomow committed:
tests: dma: add stm32h7 to loop_transfer

Add nucleo h743zi board to loop_transfer test.

Signed-off-by: Shlomi Vaknin <shlomi.39sd@gmail.com>

shlomow commented:
@FRASTM Thanks for the review! I added the board to the dma tests and it worked. The spi had some issues, but I think I solved them (it even didn't compile because some functions not available in the ll api or other things). I will open pr for this soon. I needed the nocache workaround to run the spi tests of course.

shlomow commented:
@FRASTM @erwango Can you take a look please?

manoj153 commented:
I am on stm32h7zg and I seem not to get it working properly.

 I am trying async (DMA) uart_tx, they are writing the streams to the console but they are not valid, data is messed.
nucleo-h723zg is my board in use.

I am actually suspecting `channel-config bit-mask can they be?
code tx:
```
uint8_t test_tx[] = { 0xAA, 0xBB };
uart_tx(rs485_a_dev,test_tx,2,100);
```
my device tree config:
```
&usart6 {
        pinctrl-0 = <&usart6_tx_pc6 &usart6_rx_pc7>;
        current-speed = <115200>;
        dmas =  <&dmamux1 1 72 0x440 0>, /* */
                <&dmamux1 2 71 0x480 0>;
        dma-names = "tx", "rx";
        status = "okay";
};
&dma1 {
	status = "okay";
};

&dmamux1 {
	status = "okay";
};
```
.conf :
```
CONFIG_LOG=y
CONFIG_OPENOCD_SUPPORT=y
CONFIG_PRINTK=y
CONFIG_HEAP_MEM_POOL_SIZE=256
CONFIG_ASSERT=y
CONFIG_GPIO=y
CONFIG_SERIAL=y
CONFIG_UART_STM32=y
CONFIG_UART_ASYNC_API=y
CONFIG_LOG_BACKEND_SHOW_COLOR=y
```

![image](https://user-images.githubusercontent.com/17961318/113746224-a78e0f80-9738-11eb-828d-9dea3e5e41c7.png)

shlomow commented:
Maybe the issue is with the cache. Since on h7 the data cache is enabled by default, the uart is not working properly, because the cache coherence operations are not implemented. See also #32832. I see two ways to deal with that:
1. Disable the entire data cache: in file `soc/arm/st_stm32/stm32h7/soc_m7.c` put a comment on the line `SCB_EnableDCache();`
2. Enable CONFIG_NOCACHE_MEMORY. This will create another mpu region which is not cachable, so you can place your buffer in it without being terrified that the cache will cause problems. See also the dma tests for example how to declare buffer like this.

I hope those are solving your problem. Please let me know.

manoj153 commented:
> Maybe the issue is with the cache. Since on h7 the data cache is enabled by default, the uart is not working properly, because the cache coherence operations are not implemented. See also #32832. I see two ways to deal with that:
> 
> 1. Disable the entire data cache: in file `soc/arm/st_stm32/stm32h7/soc_m7.c` put a comment on the line `SCB_EnableDCache();`
> 2. Enable CONFIG_NOCACHE_MEMORY. This will create another mpu region which is not cachable, so you can place your buffer in it without being terrified that the cache will cause problems. See also the dma tests for example how to declare buffer like this.
> 
> I hope those are solving your problem. Please let me know.

Thanks, your insights are useful. 
I will do 1 and the later if the test from 1 gives positive  results

manoj153 commented:
> 1. SCB_EnableDCache

I have taken in this PR https://github.com/zephyrproject-rtos/zephyr/pull/32832 and they fixed the data corruption issue, 
I have tested TX.

shlomow commented:
Notice that the pr you mentioned declares that not all the tests pass, so maybe you will have some problems

manoj153 commented:
> Notice that the pr you mentioned declares that not all the tests pass, so maybe you will have some problems

yeah, I am new to this project, good to learn to read the test errors. 
I will see if I can find out why the test fails.

shlomow commented:
@erwango Can you please review this PR? :)
